rfd/0156-notifications.md

```diff
@@ -0,0 +1,568 @@
+
+If this is a global notification, it will be nested in a [`GlobalNotification`](#global-notifications).
+
+```protobuf
+// Notification represents a notification item.
+message Notification {
+   // kind is the resource kind ("notification").
+   string kind = 1;
+   // sub_kind represents the unique kind of notification this is, eg. `access-request-approved`
+   string sub_kind = 2;
+   // version is the resource version.
+   string version = 3;
+   // metadata is the notification's metadata. This contains the notification's title, description, labels, and expiry.
+    teleport.header.v1.Metadata metadata = 4;
+   // spec is the notification specification.
+   NotificationSpec spec = 5;
+}
+
+// NotificationSpec is the notification specification.
+message NotificationSpec {
+   // id is the UUID of the notification, in UUIDv7 format.
+   string id = 1;
+   // created is when the notification was created, in UNIX time.
+   google.protobuf.Timestamp created = 2;
+}
+```
+
+###### User Notification State Item
+
+Each user will have a unique `NotificationState` item for every notification they have interacted with.
+
+This will keep track of whether the user has clicked on (ie. interacted with) or dismissed the notification.
+
+These will be stored per user per notification in the backend under `/notification/states/<username>/<notification id>`
+
+Dismissed states are only going to be possible for global notifications. If a user dismisses a user-specific notification directed at them, instead of updating this state to `dismissed`, the notification will just be
+deleted as well as its state item.
+
+```protobuf
+// UserNotificationState represents a notification's state for a user. This is to keep track
+// of whether the user has clicked on or dismissed the notification.
+message UserNotificationState {
+   // kind is the resource kind ("user_notification_state").
+   string kind = 1;
+   // sub_kind is the optional resource subkind. This is unused.
+   string sub_kind = 2;
+   // version is the resource version.
+   string version = 3;
+   // metadata is the user notification state's metadata.
+   teleport.header.v1.Metadata metadata = 4;
+   // spec is the user notification state's specification.
+   UserNotificationStateSpec spec = 5;
+   // status is the status of this user notification state, it contains the notification state itself which will be dynamically modified.
+   UserNotificationStateStatus status = 6;
+}
+
+// UserNotificationStateSpec is the user notification state's specification.
+message UserNotificationStateSpec {
+   // notification_id is the ID of the notification this state is for.
+   string notification_id = 1;
+}
+
+// UserNotificationSTate is the state of a notification for a user. This can represent either "clicked" or "dismissed".
+enum UserNotificationState {
+   NOTIFICATION_STATE_UNSPECIFIED = 0
+   // NOTIFICATION_STATE_UNREAD is the default state. This should never be set.
```

Comment:
An "unread" notification will have no `UserNotificationState` object. We only create it for "clicked" or "dismissed" notifications. If a notification doesn't have a state object, then it's unread. I realize now that it's confusing to have `NOTIFICATION_STATE_UNREAD` in the enum since it will never be set, I initially used that as the default but now that we have `UNSPECIFIED` it's no longer needed.